internal/hostport/hostport_nftables.go

```diff
@@ -0,0 +1,308 @@
+	if err != nil && !knftables.IsNotFound(err) {
+		return fmt.Errorf("could not list existing hostports: %w", err)
+	}
+
+	existingHairpins, err := hm.nft.ListElements(context.TODO(), "set", hairpinSet)
+	if err != nil && !knftables.IsNotFound(err) {
+		return fmt.Errorf("could not list existing hostports: %w", err)
+	}
+
+	// Delete each one that refers to this pod in its Comment.
+	tx := hm.nft.NewTransaction()
+
+	for _, elem := range existingHostPorts {
+		if elem.Comment != nil && *elem.Comment == comment {
+			tx.Delete(elem)
+		}
+	}
+
+	for _, elem := range existingHostIPPorts {
+		if elem.Comment != nil && *elem.Comment == comment {
+			tx.Delete(elem)
+		}
+	}
+
+	for _, elem := range existingHairpins {
+		if elem.Comment != nil && *elem.Comment == comment {
+			tx.Delete(elem)
+		}
+	}
+
+	if tx.NumOperations() == 0 {
+		return nil
+	}
+
+	err = hm.nft.Run(context.TODO(), tx)
+	if err != nil {
+		return fmt.Errorf("failed to clean up nftables hostport maps: %w", err)
+	}
+
+	return nil
+}
+
+// hashSandboxID hashes the sandbox ID to get a suitable identifier for an nftables
+// comment (which must be at most 128 characters).
+func hashSandboxID(id string) string {
+	hash := sha256.Sum256([]byte(id))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+
+	return encoded[:16]
+}
+
+// ensureHostPortsTable adds rules to tx to ensure the hostPortsTable is setup correctly.
+// "tx.Add" silently no-ops if the object already exists, so after the first hostport is
+// added, this function won't really do anything. (Note that for the chains (which have a
+// static set of rules), we Add+Flush the chain and then Add the desired rules, thus
+// ensuring that the chain has exactly the rules we want, while for the maps and set
+// (which have per-hostport elements), we Add them but don't modify their contents.)
+func ensureHostPortsTable(tx *knftables.Transaction, family knftables.Family) {
+	ip := "ip"
+	ipaddr := "ipv4_addr"
+
+	if family == knftables.IPv6Family {
+		ip = "ip6"
+		ipaddr = "ipv6_addr"
+	}
+
+	tx.Add(&knftables.Table{
+		Comment: knftables.PtrTo("HostPort rules created by CRI-O"),
+	})
+
+	tx.Add(&knftables.Map{
+		Name: hostPortsMap,
+		Type: knftables.Concat(
+			"inet_proto", ".", "inet_service", ":", ipaddr, ".", "inet_service",
+		),
+		Comment: knftables.PtrTo("hostports on all local IPs (protocol . hostPort -> podIP . podPort)"),
+	})
+	tx.Add(&knftables.Map{
+		Name: hostIPPortsMap,
+		Type: knftables.Concat(
+			ipaddr, ".", "inet_proto", ".", "inet_service", ":", ipaddr, ".", "inet_service",
+		),
+		Comment: knftables.PtrTo("hostports on specific IPs (hostIP . protocol . hostPort -> podIP . podPort)"),
+	})
+
+	// Create the "hostports" chain with the map lookup rules, and then create
+	// "prerouting" and "output" chains that call the "hostports" chain for
+	// locally-destined packets.
+	tx.Add(&knftables.Chain{
+		Name: "hostports",
+	})
+	tx.Flush(&knftables.Chain{
+		Name: "hostports",
+	})
```

Comment:
that is more a genuine question, is there any value on creating chains and jumping to them in nftables? 
I feel more comfortable adding all rules to the same chain , specially when tracing for debugging

Reply:
We need to run the same rules from `output` and `prerouting`, and I like having only a single copy of them.